MapboxNavigation/ImageCache.swift

```diff
@@ -0,0 +1,143 @@
+import Foundation
+
+class ImageCache: BimodalImageCache {
+
+    let memoryCache: NSCache<NSString, UIImage>
+    let diskCacheURL: URL = {
+        let fileManager = FileManager.default
+        let basePath = fileManager.urls(for: .cachesDirectory, in: .userDomainMask).first!
+        return basePath.appendingPathComponent("com.mapbox.directions.downloadedImages")
+    }()
+
+    let diskAccessQueue = DispatchQueue(label: "com.mapbox.directions.diskAccess")
+    var fileManager: FileManager?
+
+    init() {
+        memoryCache = NSCache<NSString, UIImage>()
+        memoryCache.name = "In-Memory Image Cache"
+
+        diskAccessQueue.sync {
+            fileManager = FileManager()
+        }
+
+        NotificationCenter.default.addObserver(forName: .UIApplicationDidReceiveMemoryWarning, object: nil, queue: nil) { [unowned self] (notif) in
+            self.clearMemory()
+        }
+    }
+
+    deinit {
+        NotificationCenter.default.removeObserver(self)
+    }
+
+    func store(_ image: UIImage, forKey key: String, toDisk: Bool, completion: NoArgBlock?) {
+        let key = cacheKeyForKey(key)
+
+        memoryCache.setObject(image, forKey: key as NSString, cost: cacheCostForImage(image))
+
+        let dispatchCompletion = {
+            if let completion = completion {
+                DispatchQueue.main.async {
+                    completion()
+                }
+            }
+        }
+
+        if toDisk == true {
+            diskAccessQueue.async {
+                self.createCacheDirIfNeeded()
```

Comment:
Might be mistaken but this looks like a retain cycle.

Reply:
Yes, there is a transient retain cycle there. While this block is in the queue or executing, self will be retained which ensures it sticks around long enough to complete its work. Once the block exits it is disposed of, breaking the cycle.

Does this reasoning hold water or does the team prefer a weakSelf/strongSelf pattern in these disk access methods?